### PR TITLE
Add support of Kotlin model generation

### PR DESCRIPTION
**Pull request by sav007**
Add Kotlin Codegen implemetation

Add a new concept of AST that is built from IR models, for better support  and maintenance of different targets of codegen. For now only Kotlin codegen target uses AST, Java is to be ported next.

This is PR is huge because of new generated Kotlin models for all existing Java graphql fixtures.

Part of https://github.com/apollographql/apollo-android/issues/1079

### CODE REVIEW

**Comment by bhuvan0616** [comment]
Waiting for this feature 👍

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Inflector.kt [hunk 1]
```diff
@@ -72,9 +74,8 @@ private fun irregular(): List<Pair<String, String>> {
       "move" to "moves")
 }
 
-private fun uncountable(): List<String> {
-  return listOf("equipment", "information", "rice", "money", "species", "series", "fish", "sheep")
-}
+private val uncountable = listOf("equipment", "information", "rice", "money", "species", "series", "fish", "sheep")
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Inflector.kt]: missed this earlier, what is this for?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt [hunk 2]
```diff
@@ -12,8 +12,17 @@ private val JAVA_RESERVED_WORDS = arrayOf(
     "transient", "try", "true", "void", "volatile", "while"
 )
 
+private val KOTLIN_RESERVED_WORDS = arrayOf(
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt]: I think you are missing a lot of words from https://kotlinlang.org/docs/reference/keyword-reference.html  For example  I don't see `suspend`

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt [hunk 3]
```diff
@@ -0,0 +1,121 @@
+package com.apollographql.apollo.compiler.ast
+
+data class AST(
+  val enums: List<EnumType>,
+  val customTypes: Map<String, String>,
+  val inputTypes: List<InputType>,
+  val fragments: List<FragmentType>,
+  val operations: List<OperationType>
+) {
+  data class OperationType(
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt]: any reason to have this as a nested class? We can have multiple classes in same file in kt

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    val name: String,
+    val type: Type,
+    val definition: String,
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt]: do you think we can gain readability from using type alias for some of these scalar fields?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+  sealed class FieldType {
+
+    sealed class Scalar : FieldType() {
+      object String : Scalar()
+      object Int : Scalar()
+      object Boolean : Scalar()
+      object Float : Scalar()
+      data class Enum(val typeRef: TypeRef) : Scalar()
+      data class Custom(
+        val schemaType: kotlin.String,
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt]: whats with the type?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/ASTBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,464 @@
...
+          variableName = it.variableName,
+          inverted = it.inverted
+        )
+      } ?: emptyList()
+    } else {
+      emptyList()
+    }
+  }
+
+private class Context(
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/ASTBuilder.kt]: pull up to own file? I don't see reason to keep it here

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/GraphQLKompiler.kt [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+import com.apollographql.apollo.compiler.formatPackageName
+import com.apollographql.apollo.compiler.ir.CodeGenerationIR
+import com.apollographql.apollo.compiler.ir.ScalarType
+import com.apollographql.apollo.compiler.ir.TypeDeclaration
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.moshi.Moshi
+import java.io.File
+
+class GraphQLKompiler {
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/GraphQLKompiler.kt]: cringe worthy name :-)

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/GraphQLKompiler.kt [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  data class Arguments(
+      val irFile: File,
+      val outputDir: File,
+      val customTypeMap: Map<String, String>,
+      val outputPackageName: String?,
+      val useSemanticNaming: Boolean
+  )
+}
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/GraphQLKompiler.kt]: newline?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/Utils.kt [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    is Number -> {
+      val scalarType = ScalarType.forName(graphQLType.removeSuffix("!"))
+      when (scalarType) {
+        is ScalarType.INT -> toInt()
+        is ScalarType.FLOAT -> toDouble()
+        else -> this
+      }
+    }
+    is Boolean, is Map<*, *> -> this
+    is List<*> -> this.map {
```
Comment (digitalbuddha) [inline apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/codegen/kotlin/Utils.kt]: I think you can do `mapNotNull` and not need the !!

File: gradle/dependencies.gradle [hunk 10]
```diff
@@ -31,6 +31,7 @@ ext.dep = [
     recyclerView            : "com.android.support:recyclerview-v7:$versions.supportLibVersion",
     compiletesting          : 'com.google.testing.compile:compile-testing:0.15',
     cache                   : "com.nytimes.android:cache:$versions.cacheVersion",
+    kotlinpoet              : 'com.squareup:kotlinpoet:1.0.0-RC2',
```
Comment (digitalbuddha) [inline gradle/dependencies.gradle]: I took so long to review that 1.0.0 came out :-D

Comment (alexxxdev) [reply gradle/dependencies.gradle]: it lasts so long that already 1.0.1 xD

**Comment by 0xcaff** [comment]
This seems to have gotten lost in the holiday shuffle. Do you guys need any help implementing this?

**Comment by bhuvan0616** [comment]
> This seems to have gotten lost in the holiday shuffle. Do you guys need any help implementing this?

I guess so

**Comment by alexxxdev** [comment]
@sav007 please, rebase against latest master

**Comment by friederikewild** [comment]
@sav007 Any news on solving the outstanding conflicts to get your exciting changes with the next release of Apollo?

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Inflector.kt (hunk 1 above)
Comment (sav007) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Inflector.kt]: To proper support plural / singular wording.

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt (hunk 2 above)
Comment (sav007) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/Util.kt]: Well there are 2 kind of keywords, hard and soft. This array is hard keywords, usage soft keywords depends on the context. Maybe later we need to expand this list.

File: apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt (hunk 3 above)
Comment (sav007) [reply apollo-compiler/src/main/kotlin/com/apollographql/apollo/compiler/ast/AST.kt]: There high chances to have name clash if all classes generated in the same package.

**Comment by 0xcaff** [comment]
Yay it's ready again! 🎉

**Pull request merged**